File: leetcode-practice-java/src/main/java/io/hank/leetcode/practices/graph/_994_Rotting_Oranges.java
package io.hank.leetcode.practices.graph;

import io.hank.leetcode.annotations.Topic;
import io.hank.leetcode.annotations.TopicType;
import io.hank.leetcode.practices.LeetcodeProblemSolution;

import java.util.LinkedList;
import java.util.Queue;

/**
 * <pre>
 * 在给定的 m x n 网格 grid 中，每个单元格可以有以下三个值之一：
 *
 * 值 0 代表空单元格；
 * 值 1 代表新鲜橘子；
 * 值 2 代表腐烂的橘子。
 * 每分钟，腐烂的橘子 周围 4 个方向上相邻 的新鲜橘子都会腐烂。
 *
 * 返回 直到单元格中没有新鲜橘子为止所必须经过的最小分钟数。如果不可能，返回 -1 。
 *
 *
 * 示例 1：
 *
 * 输入：grid = [[2,1,1],[1,1,0],[0,1,1]]
 * 输出：4
 *
 * 示例 2：
 *
 * 输入：grid = [[2,1,1],[0,1,1],[1,0,1]]
 * 输出：-1
 * 解释：左下角的橘子（第 2 行， 第 0 列）永远不会腐烂，因为腐烂只会发生在 4 个方向上。
 *
 * 示例 3：
 *
 * 输入：grid = [[0,2]]
 * 输出：0
 * 解释：因为 0 分钟时已经没有新鲜橘子了，所以答案就是 0 。
 *
 *
 * 提示：
 *
 * m == grid.length
 * n == grid[i].length
 * 1 <= m, n <= 10
 * grid[i][j] 仅为 0、1 或 2
 * </pre>
 */
public class _994_Rotting_Oranges extends LeetcodeProblemSolution {
    int[][] DIRECTIONS = new int[][]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    @Topic({TopicType.GRAPH, TopicType.BFS, TopicType.MATRIX})
    public int orangesRotting(int[][] grid) {
        int rows = grid.length, cols = grid[0].length;
        int freshes = 0;
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == 2) {
                    queue.offer(new int[]{i, j});
                } else if (grid[i][j] == 1) {
                    freshes++;
                }
            }
        }
        int res = 0;
        while (!queue.isEmpty()) {
            // loop all nodes in one same level
            int size = queue.size();
            boolean rotted = false;
            for (int i = 0; i < size; i++) {
                int[] index = queue.poll();
                int row = index[0], col = index[1];
                for (int[] d : DIRECTIONS) {
                    int newRow = row + d[0], newCol = col + d[1];
                    if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && grid[newRow][newCol] == 1) {
                        rotted = true;
                        queue.offer(new int[]{newRow, newCol});
                        grid[newRow][newCol] = 2;
                        freshes--;
                    }
                }
            }
            res = rotted ? res + 1 : res;
        }
        return freshes == 0 ? res : -1;
    }
}

File: leetcode-practice-typescript/src/bfs/994-rotting-oranges.ts
function orangesRotting(grid: number[][]): number {
    const m = grid.length;
    const n = grid[0].length;
    const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];
    const queue: number[][] = [];

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === 2) {
                queue.push([j, j]);
            }
        }
    }

    let steps = -1;
    while (queue.length) {
        steps++;
        const size = queue.length;
        for (let _ = 0; _ < size; _++) {
            const pos = queue.shift()!;
            const i = pos[0];
            const j = pos[1];
            for (const d of directions) {
                const ni = i + d[0];
                const nj = j + d[1];
                if (ni >= 0 && ni < m && nj >= 0 && nj < n && grid[ni][nj] === 1) {
                    // 腐烂
                    grid[ni][nj] = 2;
                    queue.push([ni, nj]);
                }
            }
        }
    }

    // 最后检查是否还有新鲜橘子
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === 1) return -1;
        }
    }
    return steps === -1 ? 0 : steps;
}

console.log(orangesRotting([[2, 1, 1], [1, 1, 0], [0, 1, 1]]));
